rust/kernel/types.rs

```diff
@@ -91,3 +91,33 @@ impl<T: PointerWrapper + Deref> PointerWrapper for Pin<T> {
         Pin::new_unchecked(T::from_pointer(p))
     }
 }
+
+/// Runs a cleanup function/closure when dropped.
+///
+/// The [`ScopeGuard::dismiss`] function prevents the cleanup function from running.
+pub struct ScopeGuard<T: FnOnce()> {
+    cleanup_func: Option<T>,
+}
+
+impl<T: FnOnce()> ScopeGuard<T> {
+    /// Creates a new cleanup object with the given cleanup function.
+    pub fn new(cleanup_func: T) -> Self {
+        Self {
+            cleanup_func: Some(cleanup_func),
+        }
+    }
+
+    /// Prevents the cleanup function from running.
+    pub fn dismiss(mut self) {
+        self.cleanup_func.take();
+    }
+}
+
+impl<T: FnOnce()> Drop for ScopeGuard<T> {
+    fn drop(&mut self) {
+        // Run the cleanup function if one is still present.
+        if let Some(cleanup) = self.cleanup_func.take() {
```

Comment:
Is the call to `take()` necessary?

Reply:
Yes, you cannot otherwise move out from a mutable reference.

Reply:
Yes. If we don't use it, we get the following error:

```
error[E0507]: cannot move out of `self.cleanup_func.0` which is behind a mutable reference
   --> rust/kernel/types.rs:119:32
    |
119 |         if let Some(cleanup) = self.cleanup_func {
    |                     -------    ^^^^^^^^^^^^^^^^^ help: consider borrowing here: `&self.cleanup_func`
    |                     |
    |                     data moved here
    |                     move occurs because `cleanup` has type `T`, which does not implement the `Copy` trait
```

If we apply the suggested fix, we get:

```
error[E0507]: cannot move out of `*cleanup` which is behind a shared reference
   --> rust/kernel/types.rs:120:13
    |
120 |             cleanup();
    |             ^^^^^^^ move occurs because `*cleanup` has type `T`, which does not implement the `Copy` trait
```

IOW, if we don't call `take`, we can't [destructure](https://doc.rust-lang.org/rust-by-example/flow_control/match/destructuring.html) the `Option<T>`.  If instead we get a reference to the `FnOnce` stored in the `Option<T>` (when there is one), then we can't call it.

Reply:
Ah of course, `T` isn't `Copy`. Thanks guys!